haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs

```diff
@@ -0,0 +1,267 @@
+--  how the corresponding value should be decoded and added to the builder.
+decodeMap ::
+    -- | Function that determines the 'MapValueDecoder' for a map key.
+    (Text -> Maybe (MapValueDecoder s builder)) ->
+    -- | Function to run the builder and produce a value.
+    (builder -> Either String r) ->
+    -- | The empty builder.
+    builder ->
+    Decoder s r
+decodeMap valDecoder runBuilder emptyBuilder = do
+    builder <-
+        decodeMapLenOrIndef >>= \case
+            Nothing -> decodeIndef emptyBuilder
+            Just len -> decodeDef len emptyBuilder
+    case runBuilder builder of
+        Left e -> fail e
+        Right r -> return r
+  where
+    decodeKV builder = do
+        key <- decodeString
+        case valDecoder key of
+            Nothing -> fail $ "Unexpected key " ++ show key
+            Just decoder -> decoder builder
+    decodeIndef builder = do
+        decodeBreakOr >>= \case
+            True -> return builder
+            False -> do
+                builder' <- decodeKV builder
+                decodeIndef builder'
+    decodeDef n builder
+        | n > 0 = decodeDef (n - 1) =<< decodeKV builder
+        | otherwise = return builder
+
+-- | Decode a CBOR decimal fraction into a 'Scientific'. The result is not normalized.
+decodeDecimalFraction :: Decoder s Scientific
+decodeDecimalFraction = do
+    tag <- decodeTag
+    unless (tag == 4) $ fail $ "Expected decimal fraction (tag 4), but found tag " ++ show tag
+    mLen <- decodeListLenOrIndef
+    forM_ mLen $ \len ->
+        unless (len == 2) $
+            fail $
+                "Expected an array of length 2, but length was " ++ show len
+    exp10 <- decodeInt
+    mantissa <- decodeInteger
+    when (isNothing mLen) $ decodeBreakOr >>= \b -> unless b (fail "Expected end of array")
+    return $ scientific mantissa exp10
+
+-- * Encoding helpers
+
+-- | An 'Encoding' that represents a key in a CBOR map. The 'Ord' instance corresponds to
+--  lexicographic ording of the CBOR encodings, so that these can be used to order keys for
+--  deterministic CBOR encoding.
+data MapKeyEncoding = MapKeyEncoding
+    { meEncoding :: Encoding,
+      meBytes :: LBS.ByteString
+    }
+
+instance Eq MapKeyEncoding where
+    (==) = (==) `on` meBytes
+
+instance Ord MapKeyEncoding where
+    compare = compare `on` meBytes
+
+-- | Convert an 'Encoding' to a 'MapKeyEncoding'
+makeMapKeyEncoding :: Encoding -> MapKeyEncoding
+makeMapKeyEncoding meEncoding = MapKeyEncoding{meBytes = CBOR.toLazyByteString meEncoding, ..}
+
+-- | Encode a map deterministically. Specifically, the map length is definite and the keys are
+--  ordered in bytewise lexicographic order. It is assumed that all encodings in the map are
+--  deterministic.
+encodeMapDeterministic :: Map.Map MapKeyEncoding Encoding -> Encoding
+encodeMapDeterministic m =
+    encodeMapLen (fromIntegral $ Map.size m)
+        <> mconcat [meEncoding k <> v | (k, v) <- Map.toAscList m]
```

Comment:
Are we doing [CDE](https://datatracker.ietf.org/doc/html/draft-ietf-cbor-cde-05) or [dCBOR](https://www.ietf.org/archive/id/draft-mcnally-deterministic-cbor-11)? The latter is a spec from BlockchainCommons which we use for address serialization, but it does not seem like there is a haskell library to do this?

Does cborg enforce "preferred serialization" when encoding?

Reply:
I think the main thing that `dCBOR` mandates beyond what we implement is NFC encoding of unicode strings. So I think we should be `CDE` compliant, and if the strings are NFC then we would also comply with `dCBOR`.

Reply:
great 😄